graphql-java-support/src/main/java/com/apollographql/federation/graphqljava/caching/CacheControlInstrumentation.java

```diff
@@ -0,0 +1,344 @@
+            .stream()
+            .map(
+                type ->
+                    CacheControlDirective.fromDirectiveContainer((GraphQLDirectiveContainer) type))
+            .filter(Optional::isPresent)
+            .forEach(directive -> fieldPolicy.restrict(directive.get()));
+      }
+    } else if (unwrappedFieldType instanceof GraphQLCompositeType) {
+
+      // Cache directive on the return type of this field if it's a composite type
+
+      Optional<CacheControlDirective> directive =
+          CacheControlDirective.fromDirectiveContainer(
+              (GraphQLDirectiveContainer) unwrappedFieldType);
+
+      if (directive.isPresent()) {
+        fieldPolicy.replace(directive.get());
+        inheritMaxAge = directive.get().getInheritMaxAge();
+      }
+    }
+
+    // Cache directive on the field itself
+
+    Optional<CacheControlDirective> fieldDirective =
+        CacheControlDirective.fromDirectiveContainer(parameters.getField());
+    if (fieldDirective.isPresent()) {
+      CacheControlDirective directive = fieldDirective.get();
+
+      // If inheritMaxAge is true, take note of that to avoid setting the
+      // default max age in the next step. This does allow setting the cache
+      // scope though.
+      //
+      // Note that specifying `@cacheControl(inheritMaxAge: true)` on a
+      // field whose return type defines a `maxAge` gives precedence to
+      // the type's `maxAge`. (Perhaps this should be some sort of
+      // error.)
+      if (directive.getInheritMaxAge() && !fieldPolicy.hasMaxAge()) {
+        inheritMaxAge = true;
+        fieldPolicy.replace(directive.getScope());
+      } else {
+        fieldPolicy.replace(directive);
+      }
+    }
+
+    // If this field returns a composite type or is a root field and
+    // we haven't seen an explicit maxAge argument, set the maxAge to 0
+    // (uncached) or the default if specified in the constructor.
+    // (Non-object fields by default are assumed to inherit their
+    // cacheability from their parents. But on the other hand, while
+    // root non-object fields can get explicit directives from their
+    // definition on the Query/Mutation object, if that doesn't exist
+    // then there's no parent field that would assign the default
+    // maxAge, so we do it here.)
+    //
+    // You can disable this on a non-root field by writing
+    // `@cacheControl(inheritMaxAge: true)` on it. If you do this,
+    // then its children will be treated like root paths, since there
+    // is no parent maxAge to inherit.
+
+    if (!fieldPolicy.hasMaxAge()
+        && ((unwrappedFieldType instanceof GraphQLCompositeType && !inheritMaxAge)
+            || parameters.getExecutionStepInfo().getPath().isRootPath())) {
+      fieldPolicy.restrict(defaultMaxAge);
+    }
+
+    return new SimpleInstrumentationContext<ExecutionResult>() {
+      @Override
+      public void onCompleted(ExecutionResult result, Throwable t) {
```

Comment:
I think I said this last time, but deferring this until the end of execution is overkill if you're not implementing dynamic cache control. On the other hand I still think implementing dynamic cache control (using a controller object that users can fetch off of their GraphQLContext that works like the JS `info.cacheControl`) wouldn't be too hard and would also allow us to remove the _Entities special-casing in this file!

Reply:
i'm interested in supporting dynamic cache control but i don't think i'll get to it in this PR. i removed the onComplete for now